File: SampSharpGameMode/Data/Extent3.cs
using SampSharp.GameMode;
using SampSharpGameMode.Extensions;
using System;
using System.Collections.Generic;
using System.Text;

namespace SampSharp.SyntheticGameMode.Data
{
    public class Extent3
    {
        public Vector3 Min { get; set; }
        public Vector3 Max { get; set; }
        private Vector3 CoordinateOrigin = new Vector3();
        public Vector3 Center => new Vector3((Min.X + Max.X) / 2, (Min.Y + Max.Y) / 2, (Min.Z + Max.Z) / 2);
        public void ChangeOrigin(Vector3 newOrigin, Vector3 rotation)
        {
            Min -= CoordinateOrigin;
            Max -= CoordinateOrigin;

            Min = Min.Rotate(rotation);
            Max = Max.Rotate(rotation);
            Min += newOrigin;
            Max += newOrigin;

            CoordinateOrigin = newOrigin;
            CoordinateOrigin = newOrigin;
        }
    }
}

File: src/utils/Extent3.lua
Extent3 = class(function(o, min, max)
    o.Min = min
    o.Max = max
    o.CoordinateOrigin = Vector3(0,0,0)
end)

function Extent3:GetCenter()
    return Vector3((self.Min.x + self.Max.x) / 2, (self.Min.y + self.Max.y) / 2, (self.Min.z + self.Max.z) / 2)
end

function Extent3:ChangeOrigin(newOrigin, rotation)
    self.Min = self.Min - CoordinateOrigin
    self.Max = self.Max - CoordinateOrigin

    self.Min = self.Min:Rotate(rotation)
    self.Max = self.Max:Rotate(rotation)
    self.Min = self.Min + newOrigin
    self.Max = self.Max + newOrigin

    self.CoordinateOrigin = newOrigin
end